sklearn/svm/src/liblinear/linear.cpp

```diff
@@ -1955,9 +1959,6 @@ static int solve_l1r_lr(
 				else
 					violation = fabs(Gn);
 
-				QP_Gmax_new = max(QP_Gmax_new, violation);
-				QP_Gnorm1_new += violation;
```

Comment:
Can we keep the code in the same location and add `QP_no_change=0` here?

Reply:
No, we cannot keep the code in the same location because it would increase the violation even though the update is skipped. This is related to the first fix, which breaks the inner loop when there is no update performed.

And no, we cannot add `QP_no_change=0` here because this variable must not be reset if there is no update in the inner loop (the condition for skipping the update is done a bit below, `if(fabs(z) < 1.0e-12)`). This is related to the second fix, which breaks the outer loop when there is no update performed in the inner-loop.